### PR TITLE
feat: bluetooth emulation

### PR DESCRIPTION
**Pull request by sadym-chromium**
And implement it for CDP.

### CODE REVIEW

File: packages/puppeteer-core/src/api/BluetoothEmulationManager.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    this.name = name;
+    this.manufacturerData = manufacturerData;
+    this.knownServiceUuids = knownServiceUuids;
+  }
+}
+
+/**
+ * @internal
+ */
+export interface BluetoothEmulationManager {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/api/BluetoothEmulationManager.ts]: does this class have to be in api?

File: packages/puppeteer-core/src/cdp/Page.ts [hunk 2]
```diff
@@ -1232,6 +1237,35 @@ export class CdpPage extends Page {
   ): Promise<DeviceRequestPrompt> {
     return await this.mainFrame().waitForDevicePrompt(options);
   }
+
+  // Use browser's singleton implementation, as Bluetooth simulation is implemented on the
+  // browser level.
+  async #cdpBluetoothEmulationManager(): Promise<CdpBluetoothEmulationManager> {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/cdp/Page.ts]: If its on the browser Puppeteer should implement it on it not the page.
This is confusing API for anyone using Puppeteer, as this will affect other pages, the way I understand it.

File: packages/puppeteer-core/src/api/BluetoothEmulationManager.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ * Copyright 2025 Google Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import type * as Bidi from 'webdriver-bidi-protocol';
+
+/**
+ * Emulated bluetooth adapter state.
+ */
+export type BluetoothAdapterState = 'absent' | 'powered-off' | 'powered-on';
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/api/BluetoothEmulationManager.ts]: All the types/interfaces/classes/etc need either `@internal` or `@public`

File: packages/puppeteer-core/src/cdp/Page.ts (hunk 2 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/Page.ts]: Web Bluetooth specification requires it to be per-page, and eventually it will be isolated to page.

File: packages/puppeteer-core/src/api/BluetoothEmulation.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
+/**
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/api/BluetoothEmulation.ts:1]: File needs to be exported from api.ts for proper documentation.

File: packages/puppeteer-core/src/cdp/Page.ts (hunk 2 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/Page.ts]: https://wpt.fyi/results/webdriver/tests/bidi/external/bluetooth/simulate_adapter/context.py?label=master&product=chrome%5Bexperimental%5D&product=firefox

File: packages/puppeteer-core/src/api/Page.ts [hunk 5]
```diff
@@ -3174,6 +3178,38 @@ export abstract class Page extends EventEmitter<PageEvents> {
...
    */
   abstract openDevTools(): Promise<Page>;
+
+  /**
+   * Disable simulated Bluetooth adapter.
+   * https://webbluetoothcg.github.io/web-bluetooth/#bluetooth-disableSimulation-command
+   *
+   * @experimental
+   */
+  abstract disableBluetoothSimulation(): Promise<void>;
```
Comment (OrKoN) [inline packages/puppeteer-core/src/api/Page.ts]: let's create a Bluetooth class and expose it via `page.bluetooth.doX()` (similar to a11y and tracing)? This would let the page API be smaller and re-use implementations between BiDi, CDP, BiDi+ more easily I think.

File: packages/puppeteer-core/src/cdp/BluetoothEmulation.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
+/**
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/cdp/BluetoothEmulation.ts:1]: Add export in `cdp.ts`.

File: packages/puppeteer-core/src/cdp/CdpBluetoothEmulation.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  // Cdp session is created on-demand.
+  #cdpSession?: CdpCDPSession;
+
+  constructor(cdpTarget: CdpTarget) {
+    this.#cdpTarget = cdpTarget;
+  }
+
+  async #getCdpSession() {
+    if (!this.#cdpSession) {
+      this.#cdpSession = await this.#cdpTarget.createCDPSession();
```
Comment (OrKoN) [inline packages/puppeteer-core/src/cdp/CdpBluetoothEmulation.ts]: why do we need to create a new session?

File: packages/puppeteer-core/src/cdp/Browser.ts [hunk 8]
```diff
@@ -483,4 +485,16 @@ export class CdpBrowser extends BrowserBase {
...
     return this.#networkEnabled;
   }
+
+  // Current Bluetooth simulation is implemented on the browser level, and not tight to
+  // the specific target. `cdpBluetoothEmulationManager` is a singleton until CDP
+  // implementation is adjusted.
+  _cdpBluetoothEmulationManager(): CdpBluetoothEmulationManager {
+    if (!this.#cdpBluetoothEmulationManager) {
+      this.#cdpBluetoothEmulationManager = new CdpBluetoothEmulationManager(
+        this.target(),
```
Comment (OrKoN) [inline packages/puppeteer-core/src/cdp/Browser.ts]: we should not base it on the deprecated target API? let's either pass the Page object or the primary client.

**Review by OrKoN — CHANGES_REQUESTED** [review]

File: packages/puppeteer-core/src/cdp/CdpBluetoothEmulation.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  BluetoothEmulationManager,
+  PreconnectedBluetoothPeripheral,
+} from '../api/BluetoothEmulationManager.js';
+
+import type {CdpCDPSession} from './CdpSession.js';
+import type {CdpTarget} from './Target.js';
+
+export class CdpBluetoothEmulationManager implements BluetoothEmulationManager {
+  #cdpTarget: CdpTarget;
+  // Cdp session is created on-demand.
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/cdp/CdpBluetoothEmulation.ts]: Why do we want to create this on demand and not use internal sessions?

File: packages/puppeteer-core/src/api/BluetoothEmulationManager.ts (hunk 1 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/api/BluetoothEmulationManager.ts]: we don't have a pattern of manager exposed from the page, so I inlined the methods there. Instead, the page can provide the manager. I don't have a strong preference here.

File: packages/puppeteer-core/src/api/BluetoothEmulationManager.ts (hunk 3 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/api/BluetoothEmulationManager.ts]: done

File: packages/puppeteer-core/src/api/Page.ts (hunk 5 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/api/Page.ts]: sure

File: packages/puppeteer-core/src/cdp/Page.ts (hunk 2 above)
Comment (Lightning00Blade) [reply packages/puppeteer-core/src/cdp/Page.ts]: Let's add that as a comment. to both the code and the `@experimental` part as this is user facing issue.

File: packages/puppeteer-core/src/api/Page.ts (hunk 5 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/api/Page.ts]: done

File: packages/puppeteer-core/src/api/BluetoothEmulationManager.ts (hunk 1 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/api/BluetoothEmulationManager.ts]: exposed it from the page

File: packages/puppeteer-core/src/cdp/Browser.ts (hunk 8 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/Browser.ts]: done

File: packages/puppeteer-core/src/cdp/CdpBluetoothEmulation.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+import type {Connection} from './Connection.js';
+
+/**
+ * @internal
+ */
+export class CdpBluetoothEmulation implements BluetoothEmulation {
+  #connection: Connection;
+
+  constructor(cdpTarget: Connection) {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/cdp/CdpBluetoothEmulation.ts]:
```suggestion
  constructor(connection: Connection) {
```

File: packages/puppeteer-core/src/api/BluetoothEmulation.ts [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+ *   ],
+ *   knownServiceUuids: ['12345678-1234-5678-9abc-def123456789'],
+ * });
+ * await page.bluetoothEmulation.disableEmulation();
+ * ```
+ *
+ * @experimental
+ * @public
+ */
+export interface BluetoothEmulation {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/api/BluetoothEmulation.ts:69]: Please add a doc line that the emulation is global but we aim at fixing it.

File: packages/puppeteer-core/src/cdp/Page.ts [hunk 12]
```diff
@@ -1232,6 +1233,12 @@ export class CdpPage extends Page {
   ): Promise<DeviceRequestPrompt> {
     return await this.mainFrame().waitForDevicePrompt(options);
   }
+
+  // Use browser's singleton implementation, as Bluetooth emulation is implemented on the
+  // browser level.
+  override get bluetoothEmulation(): BluetoothEmulation {
+    return (this.browser() as CdpBrowser)._cdpBluetoothEmulation();
```
Comment (OrKoN) [inline packages/puppeteer-core/src/cdp/Page.ts]: let's instantiate in the page constructor, use `this.#primaryTargetClient.connection()` to pass on the connection.

File: packages/puppeteer-core/src/api/Page.ts [hunk 13]
```diff
@@ -3174,6 +3175,11 @@ export abstract class Page extends EventEmitter<PageEvents> {
    * method is only available in Chrome.
    */
   abstract openDevTools(): Promise<Page>;
+
+  /**
+   * {@inheritDoc BluetoothEmulation}
+   */
+  abstract get bluetoothEmulation(): BluetoothEmulation;
```
Comment (OrKoN) [inline packages/puppeteer-core/src/api/Page.ts]: I would just call it `bluetooth` for brevity.

File: packages/puppeteer-core/src/cdp/Page.ts (hunk 2 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/Page.ts]: done

File: packages/puppeteer-core/src/api/BluetoothEmulation.ts (hunk 4 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/api/BluetoothEmulation.ts:1]: done

File: packages/puppeteer-core/src/cdp/BluetoothEmulation.ts (hunk 6 above)
Comment (OrKoN) [reply packages/puppeteer-core/src/cdp/BluetoothEmulation.ts:1]: Also I think filename should be without the Cdp prefix for consistency

File: packages/puppeteer-core/src/cdp/CdpBluetoothEmulation.ts (hunk 7 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/CdpBluetoothEmulation.ts]: removed

File: packages/puppeteer-core/src/cdp/BluetoothEmulation.ts (hunk 6 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/BluetoothEmulation.ts:1]: done

File: packages/puppeteer-core/src/api/BluetoothEmulationManager.ts (hunk 1 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/api/BluetoothEmulationManager.ts]: @OrKoN oh, I believe your concern is in the fact the file itself is placed under `api` folder. I believe its the right place, but feel free to propose any better place.

File: packages/puppeteer-core/src/cdp/CdpBluetoothEmulation.ts (hunk 9 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/CdpBluetoothEmulation.ts]: fixed

File: packages/puppeteer-core/src/cdp/CdpBluetoothEmulation.ts (hunk 10 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/CdpBluetoothEmulation.ts]: done

File: packages/puppeteer-core/src/api/BluetoothEmulation.ts (hunk 11 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/api/BluetoothEmulation.ts:69]: done

File: packages/puppeteer-core/src/cdp/Page.ts (hunk 12 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/Page.ts]: Do you know in which case `this.#primaryTargetClient.connection()` can be undefined?

File: packages/puppeteer-core/src/api/Page.ts (hunk 13 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/api/Page.ts]: done

File: packages/puppeteer-core/src/cdp/Page.ts (hunk 12 above)
Comment (OrKoN) [reply packages/puppeteer-core/src/cdp/Page.ts]: When connection is closed I believe

Comment (OrKoN) [reply packages/puppeteer-core/src/cdp/Page.ts]: Actually likely the type is wrong.

Comment (OrKoN) [reply packages/puppeteer-core/src/cdp/Page.ts]: ah sorry, it is undefined for BiDi CDP sessions

File: packages/puppeteer-core/src/api/BluetoothEmulationManager.ts (hunk 1 above)
Comment (OrKoN) [reply packages/puppeteer-core/src/api/BluetoothEmulationManager.ts]: the internal-marked BluetoothEmulationManager does not seem to exist in the latest version.

File: packages/puppeteer-core/src/cdp/Page.ts (hunk 12 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/Page.ts]: I moved the `_cdpBluetoothEmulation` to browser context.

Comment (OrKoN) [reply packages/puppeteer-core/src/cdp/Page.ts]: is there a reason to have it on the browser context? you mentioned that we will eventually have it per page. Let's instantiate it per page?

File: packages/puppeteer-core/src/api/BluetoothEmulation.ts [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+   * The manufacturer-specific data as a base64-encoded string.
+   */
+  data: string;
+}
+
+/**
+ * @public
+ * A bluetooth peripheral to be simulated.
+ */
+export class PreconnectedPeripheral {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/api/BluetoothEmulation.ts]: Do we need this to be a class? Can it be a normal object, this would be a new pattern in Puppeteer i think.

File: packages/puppeteer-core/src/cdp/BrowserContext.ts [hunk 15]
```diff
@@ -143,4 +146,11 @@ export class CdpBrowserContext extends BrowserContext {
       browserContextId: this.#id,
     });
   }
+
+  // Current Bluetooth emulation is implemented on the browser context level, and not
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/cdp/BrowserContext.ts]: Please use JSDoc comment blocks

File: packages/puppeteer-core/src/api/BluetoothEmulation.ts (hunk 14 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/api/BluetoothEmulation.ts]: We use this pattern in https://github.com/puppeteer/puppeteer/blob/79c578ed34fc8fa666047d743e152df17527b339/packages/puppeteer-core/src/api/DeviceRequestPrompt.ts#L14

Comment (sadym-chromium) [reply packages/puppeteer-core/src/api/BluetoothEmulation.ts]: Anyways, converted to interface, as the constructor is not used.

File: packages/puppeteer-core/src/api/BluetoothEmulationManager.ts (hunk 1 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/api/BluetoothEmulationManager.ts]: there was an issue in refactoring, which is addressed now.

File: packages/puppeteer-core/src/cdp/Page.ts (hunk 12 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/Page.ts]: `CdpBrowserContext` does not expose it's cdp connection. We can either expose it, or create `cdpBluetoothEmulation` per browser context. I don't have preferences on either of these options.

File: packages/puppeteer-core/src/cdp/BrowserContext.ts (hunk 15 above)
Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/BrowserContext.ts]: done

File: packages/puppeteer-core/src/cdp/Page.ts (hunk 12 above)
Comment (OrKoN) [reply packages/puppeteer-core/src/cdp/Page.ts]: I am not following:

> let's instantiate in the page constructor, use this.#primaryTargetClient.connection() to pass on the connection.

why cannot we do this as I suggested originally?

Comment (sadym-chromium) [reply packages/puppeteer-core/src/cdp/Page.ts]: Oh, we can. Done.

**Review by OrKoN — APPROVED** [review]
LGTM thanks!

**Comment by sadym-chromium** [comment]
@Lightning00Blade do you have other concerns?

File: test/src/bluetooth-emulation.spec.ts [hunk 16]
```diff
@@ -29,30 +28,12 @@ describe('device request prompt with emulated device', function () {
...
-    );
-  });
-
-  afterEach(async () => {
-    await browserSession?.detach();
-  });
-
   it('can be canceled', async function () {
     const {page} = state;
+    await state.page.goto(state.httpsServer.EMPTY_PAGE);
```
Comment (Lightning00Blade) [inline test/src/bluetooth-emulation.spec.ts]: Nit: use page as it's destructed already

**Review by Lightning00Blade — APPROVED** [review]

File: test/src/bluetooth-emulation.spec.ts (hunk 16 above)
Comment (sadym-chromium) [reply test/src/bluetooth-emulation.spec.ts]: done, thanks!
